test/rust_toolchain_template_variable_info/defs.bzl

```diff
@@ -0,0 +1,26 @@
+"""A module which defines rules for testing TemplateVariableInfo"""
```

Comment:
There's already similar tests in [//test/unit/toolchain_make_variables/toolchain_make_variables_test.bzl](https://github.com/bazelbuild/rules_rust/blob/0.6.0/test/unit/toolchain_make_variables/toolchain_make_variables_test.bzl), could you instead expand the pattern there to cover the new cases?

Reply:
Ah, yes. Will merge them. Thanks!

Reply:
Only trouble is https://github.com/bazelbuild/bazel-skylib/issues/375 sort of blocks it.

Reply:
Do you need to write a rule that consumes the toolchain? Is the target you wrote for the test the thing you want to do? Have some target with:

```starlark
env = {
    "CARGO": "$(CARGO)",
    "RUSTC": "$(RUSTC)",
}
```

That should be testable in a similar way to the current tests. You're not transitioning these targets so couldn't you also use `current_rust_toolchain` to find values to match against if you wanted to test environment variables for the action?

Reply:
What do you two think of the testing on https://github.com/bazelbuild/rules_rust/pull/1416 ?

Reply:
LGTM!

Reply:
Looks okay, though there should be a test for the "toolchains" attribute directly on the rule. This should also enable variable substitution.

Reply:
Yes, specifically it need to be part of the rule definition itself, not a target. Such:

```starlark
_ = rule(
    implementation = ...,
    toolchains = ["//rust/toolchain"],
)
```

Reply:
Ah, isn’t that currently covered by the existing rust rules? Either way, it would be easy to write an new rule to cover that.

Reply:
>isn’t that currently covered by the existing rust rules?

I don't think so. We're specifically testing for the usage of template variables when using the toolchain.

>Either way, it would be easy to write a new rule to cover that.

I think so too :)

Reply:
Even better, some usages of `"$(rootpath @rules_rust//rust/toolchain:current_cargo_files)"` should be replaced with the template variable info.

https://cs.github.com/bazelbuild/rules_rust?q=%22%24%28rootpath+%40rules_rust%2F%2Frust%2Ftoolchain%3Acurrent_cargo_files%29%22

https://github.com/bazelbuild/rules_rust/blob/16175c881cc1070c13a57c8db9d73ca2cab39acb/crate_universe/BUILD.bazel#L105

https://github.com/bazelbuild/rules_rust/blob/16175c881cc1070c13a57c8db9d73ca2cab39acb/crate_universe/tools/cross_installer/BUILD.bazel#L22

https://github.com/bazelbuild/rules_rust/blob/16175c881cc1070c13a57c8db9d73ca2cab39acb/crate_universe/test_data/private/BUILD.bazel#L11

Reply:
https://github.com/bazelbuild/rules_rust/pull/1412#discussion_r900284602 I think this should be a separate PR but sounds reasonable.

@uhthomas Can you take another look? I've updated the tests.

Reply:
Seems like the functionality you want is only supported in Bazel >5?
https://storage.googleapis.com/bazel-untrusted-buildkite-artifacts/0181729b-eb6e-47cf-8dc8-e291aac72a12/test/unit/toolchain_make_variables/rust_toolchain_make_variable_expansion_test/test.log

Reply:
No? It outputs a path just fine, though the path seems to be different?

```
In test _rust_toolchain_make_variable_expansion_test from //test/unit/toolchain_make_variables:toolchain_make_variables_test.bzl: Expected env[ENV_VAR_CARGO] to be equal to 'bazel-out/k8-fastbuild/bin/external/rust_linux_x86_64/toolchain_for_x86_64-unknown-linux-gnu_impl/bin/cargo', got 'bazel-out/host/bin/external/rust_linux_x86_64/toolchain_for_x86_64-unknown-linux-gnu_impl/bin/cargo'
```

```diff
-bazel-out/k8-fastbuild/bin/external/rust_linux_x86_64/toolchain_for_x86_64-unknown-linux-gnu_impl/bin/cargo|
+bazel-out/host/bin/external/rust_linux_x86_64/toolchain_for_x86_64-unknown-linux-gnu_impl/bin/cargo
````

Reply:
Is this something you'd be able to pickup and find a solution for? It's also odd to me that the tests passed on all other jobs but not the Min Bazel Version ones